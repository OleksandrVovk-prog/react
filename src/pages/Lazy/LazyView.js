import { lazy, Suspense } from 'react';
import PropTypes from 'prop-types';
import Button from '../../components/Button/Button';
import styles from './sass/Lazy.module.scss';

const Film = lazy(() => import('../../components/Film/Film'));

export default function LazyView({
  information, getFilm, title, buttonTitle, loading, search, handleSearch
}) {
  return (
    <div>
      <p className={styles.appIntro}>{title}</p>
      <Suspense fallback={<p>Loading component...</p>}>
        {
          loading
            ? <p>Loading...</p>
            : information.Title && <Film film={information} />
        }
      </Suspense>
      <input type="text" value={search} onChange={handleSearch} data-testid="input-search" />
      <Button
        title={buttonTitle}
        buttonClass={styles.updateFilm}
        handleClick={getFilm}
        id="update-film-button"
      />
    </div>
  );
}

LazyView.propTypes = {
  information: PropTypes.shape({
    Title: PropTypes.string
  }),
  getFilm: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  buttonTitle: PropTypes.string.isRequired,
  loading: PropTypes.bool.isRequired,
  search: PropTypes.string.isRequired,
  handleSearch: PropTypes.func.isRequired,
};

LazyView.defaultProps = {
  information: {}
};
