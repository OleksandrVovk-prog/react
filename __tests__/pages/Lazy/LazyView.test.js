import React from 'react';
import userEvent from '@testing-library/user-event';
import i18next from 'i18next';
import { act } from '@testing-library/react';
import { render } from '../../settings/test-utils';
import LazyView from '../../../src/pages/Lazy/LazyView';

describe('<LazyView />', () => {
  let wrapper;
  const mockFunc = jest.fn();

  const props = {
    information: {
      Title: 'mock information title',
      Poster: 'mock information poster'
    },
    getFilm: mockFunc,
    title: i18next.t('lazyPage'),
    buttonTitle: i18next.t('search'),
    loading: false,
    search: 'mock search',
    handleSearch: mockFunc
  };

  beforeEach(async () => {
    wrapper = await act(async () => render(
      <LazyView
        information={props.information}
        getFilm={props.getFilm}
        title={props.title}
        buttonTitle={props.buttonTitle}
        loading={props.loading}
        search={props.search}
        handleSearch={props.handleSearch}
      />
    ));
  });

  test('should render LazyView component', () => {
    const { asFragment } = wrapper;
    expect(asFragment(wrapper)).toMatchSnapshot();
  });

  test('should render Button component', () => {
    const { getByTestId } = wrapper;
    const Button = getByTestId('update-film-button');
    expect(Button).toBeInTheDocument();
  });

  test('should render loading element', async () => {
    const { getByText } = await act(async () => render(<LazyView
      information={props.information}
      getFilm={props.getFilm}
      title={props.title}
      buttonTitle={props.buttonTitle}
      loading
      search={props.search}
      handleSearch={props.handleSearch}
    />));
    const Loading = getByText(/loading/i);
    expect(Loading).toBeInTheDocument();
  });

  test('on Button click should call mockFunc function', async () => {
    const { getByTestId } = wrapper;
    const Button = getByTestId('update-film-button');
    await userEvent.click(Button);
    expect(mockFunc).toHaveBeenCalled();
  });
});
