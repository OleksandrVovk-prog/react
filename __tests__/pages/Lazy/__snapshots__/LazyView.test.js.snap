// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LazyView /> should render LazyView component 1`] = `
<DocumentFragment>
  <div>
    <p
      class="appIntro"
    >
      Lazy page
    </p>
    <div>
      <p>
        mock information title
      </p>
      <img
        alt="mock information title"
        class="appBanner"
        src="mock information poster"
      />
    </div>
    <input
      data-testid="input-search"
      type="text"
      value="mock search"
    />
    <button
      class="updateFilm"
      data-testid="update-film-button"
      type="button"
    >
      Search
    </button>
  </div>
</DocumentFragment>
`;
